src/wallet/wallet.h

```diff
@@ -278,14 +289,11 @@ struct CNotePlaintextEntry
     libzcash::NotePlaintext plaintext;
 };
 
-
-
-/** A transaction with a merkle branch linking it to the block chain. */
-class CMerkleTx : public CTransaction
+class MerkleAbstractBase : virtual public CTransactionBase
```

Comment:
Let's collapse MerkleAbstractBase, and CWalletObj and rename it to CWalletTransactionBase.
Remove CMerkleTx and CMerkleCert and move the serialization code directly in CWalletCert (that extends CWalletTransactionBase) and CWalletTransaction (extending CWalletTransactionBase).
Introduce a pure virtual IsOutputMatureAtHeight(int n, int height) method in CWalletTransactionBase.
Store in CWalletCert and CWalletTransaction all the needed informations to calculate the maturity of each output and all the information needed to not to have to access the Coins database. Regarding this it will be also necessary to mark a certificate as ceased if a sidechain has been ceased at a certain block. Also this information should be part of CWalletCert. We should also have to introduce two new signals raised by main.cpp in case of SidechainCeased and SidechainResurrected (in case of block revert). Both events are passing the sidechainId and the lastCertificateId.

Reply:
Collapsed MerkleAbstractBase and CWalletObj  into CWalletTransactionBase; also removed inheritance from CTransactionBase, hence no diamond inheritance for CWalletTx/Certs.  Removed CMerkleTx and CMerkleCert.